### ISSUE TITLE
Vector layers are cropped when rotated

### ISSUE THREAD

**Issue by fnicollet** [hyperknot/Leaflet#1]
To reproduce the issue, use map.html, zoom 5x, click 30 to change the bearing, slightly move the map, the vector will be cropped:
![image](https://cloud.githubusercontent.com/assets/1240481/16100790/2ab78bd4-3361-11e6-9634-935146c9409c.png)
Looks like the vector is cropped to the non-rotated extent, most probably in this L.Polyline code where the segments are clipped:
https://github.com/hyperknot/Leaflet/blob/debug/src/layer/vector/Polyline.js#L168

**Comment by fnicollet** [comment]
I am unsure on how to fix this. 

When rotate is on, all the panes belong to the rotatepan. The rotatepan is rotated and so all the pans are naturally rotatedas their are relative to its  position. For tile images, it's not really an issue as they are "absolute" to their container, they are just placed with 3d transform. The tile container has no boundaries and tiles can be placed basically anywhere.

In the overlay pane, there is a canvas, which has a fixed width/height. Anything drawn out of its boundaries will be cropped. So if you want anything to be displayed in the canvas, it needs to cover *at least* the whole viewport. If you rotate the canvas, well, you need a much larger canvas than just the rectangle of the map div as the canvas cannot have a distorted form, it needs to be rectangle. 

So the canvas would need to be a lot bigger and translated so that it covers the whole viewport. I don't think the clipping algorithm would work well too as it needs some real bounds, not distorted bounds, so you would end up drawing much more than needed and the map div would crop that out.

Would it be better to keep the overlay pane outside of the rotate pane? so that it is still the size of the viewport and then, the *content* of the canvas would need to be drawn as rotated

What do you think?

**Comment by IvanSanchez** [comment]
This must be an issue similar to the tile layers visible bounds. For that, I think I just added a lot of padding (+40% - +50%) to compensate. Leaflet crops the renderers (the `<canvas>` or the `<svg>`) in a similar fashion.

Either recalculate the cropping area taking rotation into account, or just add padding like crazy.

**Comment by fnicollet** [comment]
Thanks @IvanSanchez :)

I found a way to make it actually work by keeping the overlayPane out of the rotatePane and keeping it straight, then when drawing, account for possible rotation. Works pretty good, I even managed to keep the transitions working while zooming.
I have one last fix I need to do about some cropping which doesn't account for rotation yet, then i'll submit a PR (currently sitting in https://github.com/fnicollet/Leaflet/tree/patch-3)

**Comment by hyperknot** [comment]
@fnicollet Thanks for looking into this and preparing PRs. This one might be straightforward to port over from @IvanSanchez's changes.

Here is a squash from all of Ivan's work on the old rotate branch:
https://github.com/hyperknot/Leaflet/commit/0448ffd8f28730ff3c59822351d4a04f54a3972f

I'm porting these one by one, spending some time to on making them as clean and reliable as I can and I have yet to touch the vector rendering code. So feel free to attach those changes to your PR!

Also, you can make smaller PRs, it's easier to decide on small changes.

**Comment by fnicollet** [comment]
Ah I thought most @IvanSanchez work were back ported into this branch apart from the touch gesture. Should have looked before, would have saved me some headaches :)
@IvanSanchez solution is different (overlayPane is also rotated but with a lot of padding to account for rotation. Code is better than mine and impact less files, which probably also means less regressions on his side. 

@hyperknot Can you cherry pick his commits on Canvas.js and Renderer.js and merge them with your branch please?

Thanks,

**Comment by hyperknot** [comment]
@fnicollet it is not possible to cherry pick commits from the old branch, there has been thousands of commits since then. Only manually implementing them is possible, making sure that nothing brakes.

**Comment by fnicollet** [comment]
Getting to it in https://github.com/fnicollet/Leaflet/tree/rotate-rc1 but I am running into issues with backporting this fix from @IvanSanchez https://github.com/Leaflet/Leaflet/commit/a2595bbfe8109dc9cd24a15d28e0e4e6048e2e5f
Transformation is now done using the center and not the topLeft corner which was clamped (calculated in lat/lng). 
I am a bit lost in rotations/translations and can't figure out how to fix it. There must be a couple of rotateFrom() instructions missing here and there but I can't find it yet ...

**Comment by hyperknot** [comment]
I'll join you on this in the next days, I hope together we can fix it properly.

**Comment by fnicollet** [comment]
Last big issue of the rotate-rc1 branch before it is actually useable in most cases. Solution must be so close, yet so hard to find :)

**Comment by hyperknot** [comment]
Yes, I'll look into it properly and together we can definitely make it work well.

Also, we can take it a bit easy and not support old Android browsers and old IE if that helps a bit.
